### ISSUE TITLE
App crashes when distributed or exported as .IPA

### ISSUE THREAD

**Issue by cennydavidsson** [52inc/Pulley#172]
I have tried Pulley with both Carthage and with manual installation, and with IB or with code with the same result. If I load the app on to an iPhone through Xcode everything works fine. If I then later use the same code and exported it to an .IPA and load it via iTunes or iTunesConnects TestFlight it crashes as soon as the Pulley view loads. You can find the stack trace below.

Last Exception Backtrace:
0   CoreFoundation                	0x183f57164 __exceptionPreprocess + 124
1   libobjc.A.dylib               	0x1831a0528 objc_exception_throw + 55
2   CoreFoundation                	0x183f570ac +[NSException raise:format:] + 115
3   UIKit                         	0x18dc52244 UINibDecoderDecodeObjectForValue + 371
4   UIKit                         	0x18dc524f8 UINibDecoderDecodeObjectForValue + 1063
5   UIKit                         	0x18dc520b8 -[UINibDecoder decodeObjectForKey:] + 103
6   UIKit                         	0x18d7df74c -[UIView initWithCoder:] + 651
7   UIKit                         	0x18dc52380 UINibDecoderDecodeObjectForValue + 687
8   UIKit                         	0x18dc520b8 -[UINibDecoder decodeObjectForKey:] + 103
9   UIKit                         	0x18daa46c8 -[UIRuntimeConnection initWithCoder:] + 187
10  UIKit                         	0x18dc52380 UINibDecoderDecodeObjectForValue + 687
11  UIKit                         	0x18dc524f8 UINibDecoderDecodeObjectForValue + 1063
12  UIKit                         	0x18dc520b8 -[UINibDecoder decodeObjectForKey:] + 103
13  UIKit                         	0x18daa3a08 -[UINib instantiateWithOwner:options:] + 1167
14  UIKit                         	0x18d880be0 -[UIViewController _loadViewFromNibNamed:bundle:] + 371
15  UIKit                         	0x18d62e8e8 -[UIViewController loadView] + 175
16  UIKit                         	0x18d5116a8 -[UIViewController loadViewIfRequired] + 183
17  UIKit                         	0x18d5115d8 -[UIViewController view] + 27
18  ***-prototyp             	0x1046ba2fc _hidden#838_ + 95
19  ***-prototyp             	0x1046b0984 _hidden#739_ + 51588 (__hidden#372_:0)
20  ***-prototyp             	0x1046c2934 _hidden#1884_ + 125236 (__hidden#1907_:0)
21  ***-prototyp             	0x1046c2a68 _hidden#1885_ + 125544 (__hidden#1907_:0)
22  UIKit                         	0x18d511a00 -[UIViewController loadViewIfRequired] + 1039
23  UIKit                         	0x18d5115d8 -[UIViewController view] + 27
24  UIKit                         	0x18d633878 -[UITabBarController transitionFromViewController:toViewController:transition:shouldSetSelected:] + 767
25  UIKit                         	0x18d633564 -[UITabBarController transitionFromViewController:toViewController:] + 67
26  UIKit                         	0x18d6333c0 -[UITabBarController _setSelectedViewController:] + 399
27  UIKit                         	0x18d635938 -[UITabBarController setSelectedViewController:] + 111
28  UIKit                         	0x18d6e93c8 -[UITabBarController _tabBarItemClicked:] + 611
29  UIKit                         	0x18d5465cc -[UIApplication sendAction:to:from:forEvent:] + 95
30  UIKit                         	0x18d6e90f8 -[UITabBar _sendAction:withEvent:] + 527
31  UIKit                         	0x18d5465cc -[UIApplication sendAction:to:from:forEvent:] + 95
32  UIKit                         	0x18d54654c -[UIControl sendAction:to:forEvent:] + 79
33  UIKit                         	0x18d5310f4 -[UIControl _sendActionsForEvents:withEvent:] + 439
34  UIKit                         	0x18dab580c -[UITabBar _buttonUp:] + 119
35  UIKit                         	0x18d5465cc -[UIApplication sendAction:to:from:forEvent:] + 95
36  UIKit                         	0x18d54654c -[UIControl sendAction:to:forEvent:] + 79
37  UIKit                         	0x18d5310f4 -[UIControl _sendActionsForEvents:withEvent:] + 439
38  UIKit                         	0x18d545e40 -[UIControl touchesEnded:withEvent:] + 575
39  UIKit                         	0x18d545960 -[UIWindow _sendTouchesForEvent:] + 2543
40  UIKit                         	0x18d540e78 -[UIWindow sendEvent:] + 3207
41  UIKit                         	0x18d511e7c -[UIApplication sendEvent:] + 339
42  UIKit                         	0x18de6730c __dispatchPreprocessedEventFromEventQueue + 2363
43  UIKit                         	0x18de69898 __handleEventQueueInternal + 4759
44  UIKit                         	0x18de627b0 __handleHIDEventFetcherDrain + 151
45  CoreFoundation                	0x183eff77c __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 23
46  CoreFoundation                	0x183eff6fc __CFRunLoopDoSource0 + 87
47  CoreFoundation                	0x183efef84 __CFRunLoopDoSources0 + 203
48  CoreFoundation                	0x183efcb5c __CFRunLoopRun + 1047
49  CoreFoundation                	0x183e1cc58 CFRunLoopRunSpecific + 435
50  GraphicsServices              	0x185cc8f84 GSEventRunModal + 99
51  UIKit                         	0x18d5755c4 UIApplicationMain + 235
52  ***-prototyp             	0x1046ab5bc main + 30140 (__hidden#292_:6)
53  libdyld.dylib                 	0x18393c56c start + 3

**Comment by brendan09** [comment]
The Stacktrace doesn’t provide enough information to tell you the exact error. However, it looks like it’s having an issue trying to load your XIB (or Storyboard) file. 

Make sure you don’t have any IBOutlets connected to variables that no longer exist (or who’s name has changed) and that your release configuration / target is linking Pulley and has any module specifications in IB set correctly. If you’re using CocoaPods or Carthage you’ll need to make sure you specify Pulley as the module for the view controller. Make sure Pulley is added to your build target, as a framework or as files that go in the build for the target.

The issue isn’t related to Pulley, but rather with the way you have IB setup to load Pulley. It’s not able to find something, either an outlet to connect (which indicates connected to something that no longeto exists) or the class itself (which means it didn’t get linked in your build process). 

Without more information on your setup, I’m not able to help locate your error exactly. Those are some common places people have issues with IB.

If you have a project or sample project with the issue and would like me to take a look, feel free to email it to me: brendan@52inc.com

**Comment by brendan09** [comment]
@cennydavidsson I took a look at your project and did some experimenting. I have some info that _may_ help you track down the issue.

1. Because you're not creating PulleyViewController in Interface Builder, this stack trace / crash isn't caused by Pulley. This stack trace is occurring while attempting to load a view controller from a Storyboard.

2. As a result of the above item, this means that your crash is coming from an error trying to load _something_ in one of the view controllers that you're instantiating from the storyboard to use with Pulley. The crash occurs _before_ the PulleyViewController is created.

3. My best guess is a Carthage framework missing in release builds, and the missing framework is causing a crash. I saw that you had an MXSegmentedControl framework trying to be copied by a build script for Carthage. However, you never actually added that framework to the "Linked Frameworks and Libraries" for your target. Although it would find it and run in debug, it wouldn't be copied into the final release bundle and would therefore crash if you ever tried to instantiate that object type (or any from that framework).

However, because this crash is occurring _before_ PulleyViewController is created, indicating this isn't a problem with Pulley, there's unfortunately not much I can do to help you. I'd recommend checking your Carthage frameworks are setup correctly, and then looking close to see what would cause your Storyboard view controllers to fail during instantiation from the Storyboard.
